### PR TITLE
feat(conditions): add relative timestamp datatype

### PR DESCRIPTION
**Pull request by sileht**
## feat(conditions): add relative timestamp datatype

This change adds a new datatype for some conditions:

14 days ago, to defined a relative timestamp to the current date and
time.

Change-Id: I625efef7959b7852196691947c57c01d7de12851

## fix(conditions): rename current-datetime in current-timestamp

Change-Id: Iaf2e6a43357b00f5e4352671cb1d244efd0f4dfe

### CODE REVIEW

File: docs/source/configuration.rst [hunk 1]
```diff
@@ -197,6 +197,37 @@ Examples
...
               The pull request base branch has reached end-of-life.
 
+
+.. _relative datetime:
+
+Relative datetime
+~~~~~~~~~~~~~~~~~
+
+A datetime can be express as relative of the current date and time.
+The format is "[DD days] [HH:MM] ago":
```
Comment (jd) [inline docs/source/configuration.rst:207]:
```suggestion
The format is ``[DD days] [HH:MM] ago``:
```

File: docs/source/configuration.rst [hunk 2]
```diff
@@ -197,6 +197,37 @@ Examples
...
+
+* DD, the number of days
+* HH, the number of hours
+* MM, the number of minutes
+
+If the current date is 18 June 2020, `updated-at>=14 days ago` will be translated `updated-at>=2020-06-04T00:00:00`.
```
Comment (jd) [inline docs/source/configuration.rst:213]:
```suggestion
If the current date is 18th June 2020, `updated-at>=14 days ago` will be translated `updated-at>=2020-06-04T00:00:00`.
```

File: docs/source/configuration.rst [hunk 3]
```diff
@@ -197,6 +197,37 @@ Examples
...
+
+      - name: close stale pull request
+        conditions:
+          - base=main
+          - -closed
+          - updated-at>=14 days ago
+        actions:
+          close:
+            message: |
+              This pull request looks stale. Feel free to reopen it if you thinkg it's a mistake.
```
Comment (jd) [inline docs/source/configuration.rst:228]:
```suggestion
              This pull request looks stale. Feel free to reopen it if you think it's a mistake.
```

File: docs/source/configuration.rst [hunk 4]
```diff
@@ -197,6 +197,37 @@ Examples
             message: |
               The pull request base branch has reached end-of-life.
 
+
+.. _relative datetime:
+
+Relative datetime
```
Comment (jd) [inline docs/source/configuration.rst:203]:
```suggestion
.. _relative timestamp:

Relative timestamp
```

Datetime is not a word, timestamp is.

We should also fix the previous type, my bad.

File: mergify_engine/rules/parser.py [hunk 5]
```diff
@@ -95,6 +96,29 @@
...
+    r"(?P<hours>\d+):"
+    r"(?P<minutes>\d\d)"
+    r")? ago$"
+)
+
+
+def _parse_timedelta_to_now(tokens):
+    m = _TIMEDELTA_TO_NOW_RE.match(tokens[0])
+    if m is None:
+        raise pyparsing.ParseException("invalid time interval")
```
Comment (jd) [inline mergify_engine/rules/parser.py:113]:
```suggestion
        raise pyparsing.ParseException("invalid relative timestamp")
```

**Review by jd — DISMISSED** [review]

File: docs/source/configuration.rst [hunk 6]
```diff
@@ -168,12 +168,12 @@ Examples
           merge:
             method: merge
 
-.. _iso datetime:
+.. _iso timestamp:
 
-Datetime
-~~~~~~~~
+timestamp
```
Comment (jd) [inline docs/source/configuration.rst:173]:
```suggestion
Timestamp
```

File: docs/source/configuration.rst [hunk 7]
```diff
@@ -168,12 +168,12 @@ Examples
...
+~~~~~~~~~
 
-The date and datetime format must follow the ISO 8601 standard. If timezone is missing, it's assumed to be UTC.
+The date and timestamp format must follow the ISO 8601 standard. If timezone is missing, it's assumed to be UTC.
```
Comment (jd) [inline docs/source/configuration.rst:176]:
```suggestion
The timestamp format must follow the ISO 8601 standard. If the timezone is missing, the timestamp is assumed to be in UTC.
```

Add a link to https://en.wikipedia.org/wiki/ISO_8601

File: docs/source/configuration.rst [hunk 8]
```diff
@@ -197,6 +197,37 @@ Examples
             message: |
               The pull request base branch has reached end-of-life.
 
+
+.. _relative timestamp:
+
+Relative timestamp
```
Comment (jd) [inline docs/source/configuration.rst:203]:
```suggestion
Relative Timestamp
```

File: docs/source/configuration.rst [hunk 9]
```diff
@@ -197,6 +197,37 @@ Examples
...
+~~~~~~~~~~~~~~~~~~
+
+A datetime can be express as relative of the current date and time.
```
Comment (jd) [inline docs/source/configuration.rst:206]:
```suggestion
Timestamps can be expressed relative to the current date and time.
```

File: docs/source/configuration.rst [hunk 10]
```diff
@@ -197,6 +197,37 @@ Examples
...
+The format is ``[DD days] [HH:MM] ago``:
+
+* DD, the number of days
+* HH, the number of hours
+* MM, the number of minutes
+
+If the current date is 18th June 2020, `updated-at>=14 days ago` will be translated `updated-at>=2020-06-04T00:00:00`.
```
Comment (jd) [inline docs/source/configuration.rst:213]:
```suggestion
If the current date is 18th June 2020, ``updated-at>=14 days ago`` will be translated ``updated-at>=2020-06-04T00:00:00``.
```

File: docs/source/spelling_wordlist.txt [hunk 11]
```diff
@@ -25,3 +25,4 @@ regexes
 Snyk
 templating
 wasn
+th
```
Comment (jd) [inline docs/source/spelling_wordlist.txt:28]: why?

**Review by jd — CHANGES_REQUESTED** [review]

File: docs/source/conditions.rst [hunk 12]
```diff
@@ -254,19 +254,19 @@ Here's the list of pull request attribute that can be used in conditions:
...
-     - :ref:`Datetime <iso datetime>`
+     - :ref:`Timestamp <iso timestamp>` or :ref:`Relative timestamp <relative timestamp>`
      - The time the pull request was merged at.
    * - ``closed-at``
-     - :ref:`Datetime <iso datetime>`
+     - :ref:`Timestamp <iso timestamp>` or :ref:`Relative timestamp <relative timestamp>`
      - The time the pull request was closed at.
-   * - ``current-datetime``
-     - :ref:`Datetime <iso datetime>`
+   * - ``current-timestamp``
```
Comment (jd) [inline docs/source/conditions.rst:268]: did the code change for this?

File: docs/source/spelling_wordlist.txt (hunk 11 above)
Comment (sileht) [reply docs/source/spelling_wordlist.txt:28]: spelling complains for th in 13th

File: docs/source/conditions.rst (hunk 12 above)
Comment (sileht) [reply docs/source/conditions.rst:268]: now yes

File: docs/source/configuration.rst [hunk 13]
```diff
@@ -197,6 +199,37 @@ Examples
...
+Examples
+++++++++
+
+.. code-block:: yaml
+
+      - name: close stale pull request
+        conditions:
+          - base=main
+          - -closed
+          - updated-at>=14 days ago
```
Comment (jd) [inline docs/source/configuration.rst:226]:
```suggestion
          - updated-at<14 days ago
```

Funny that you actually made the mistake 😆

**Review by jd — DISMISSED** [review]

File: docs/source/configuration.rst (hunk 13 above)
Comment (sileht) [reply docs/source/configuration.rst:226]: I copy/paste the other PR that have the wrong implementation.

**Review by jd — APPROVED** [review]

**Pull request merged**